vsphere/resource_vsphere_vmfs_datastore.go

```diff
@@ -0,0 +1,324 @@
+	}
+
+	id := d.Id()
+	ds, err := datastoreFromID(client, id)
+	if err != nil {
+		return fmt.Errorf("cannot find datastore: %s", err)
+	}
+
+	// Veto this update if it means a disk was removed. Shrinking
+	// datastores/removing extents is not supported.
+	old, new := d.GetChange("disks")
+	for _, v1 := range old.([]interface{}) {
+		var found bool
+		for _, v2 := range new.([]interface{}) {
+			if v1.(string) == v2.(string) {
+				found = true
+			}
+		}
+		if !found {
+			return fmt.Errorf("disk %s found in state but not config (removal of disks is not supported)", v1)
+		}
+	}
+
+	// Maintain a copy of the disks that have been added just in case the update
+	// fails in the middle.
+	old2 := make([]interface{}, len(old.([]interface{})))
+	copy(old2, old.([]interface{}))
+
+	// Now we basically reverse what we did above when we were checking for
+	// removed disks, and add any new disks that have been added.
+	for _, v1 := range new.([]interface{}) {
+		var found bool
+		for _, v2 := range old.([]interface{}) {
+			if v1.(string) == v2.(string) {
+				found = true
+			}
+		}
+		if !found {
+			// Add the disk
+			spec, err := diskSpecForExtend(dss, ds, v1.(string))
+			if err != nil {
+				return err
+			}
+			ctx, cancel := context.WithTimeout(context.Background(), defaultAPITimeout)
+			defer cancel()
+			if _, err := extendVmfsDatastore(ctx, dss, ds, *spec); err != nil {
+				return err
+			}
+			// The add was successful. Since this update is not atomic, we need to at
+			// least update the disk list to make sure we don't attempt to add the
+			// same disk twice.
+			old2 = append(old2, v1)
+			if err := d.Set("disks", old2); err != nil {
+				return fmt.Errorf(formatUpdateInconsistentState, v1.(string), err)
+			}
+		}
+	}
+
+	// Should be done with the update here.
+	return resourceVSphereVmfsDatastoreRead(d, meta)
+}
+
+func resourceVSphereVmfsDatastoreDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*govmomi.Client)
+	hsID := d.Get("host_system_id").(string)
+	dss, err := hostDatastoreSystemFromHostSystemID(client, hsID)
+	if err != nil {
+		return fmt.Errorf("error loading host datastore system: %s", err)
+	}
+
+	id := d.Id()
+	ds, err := datastoreFromID(client, id)
+	if err != nil {
+		return fmt.Errorf("cannot find datastore: %s", err)
+	}
+
+	// This is a race that more than likely will only come up during tests, but
+	// we still want to guard against it - when working with datastores that end
+	// up mounting across multiple hosts, removing the datastore will fail if
+	// it's removed too quickly (like right away, for example). So we set up a
+	// very short retry waiter to make sure if the first attempt fails, the
+	// second one should probably succeed right away. We also insert a small
+	// minimum delay to make an honest first attempt at trying to delete the
+	// datastore without spamming the task log with errors.
+	deleteRetryFunc := func() (interface{}, string, error) {
```

Comment:
again, this feels like SDK logic

Reply:
There is a property waiter, but it's unclear what property to follow here - I think the major factor in this race is if the disk is a shared VMFS disk or not (example: iSCSI).

This only started coming up when I started testing iSCSI VMFS disks versus local ones (local disks actually have another issue where you can't add extents through vCenter, you need to connect to the host directly, so I moved the vCenter tests over to use iSCSI disks that I provisioned from the lab NAS).

In the event of a shared VMFS datastore, vCenter will automatically try to mount it on every host in a cluster that can access it, and I think that part is causing issues - the task returns, yet other hosts are still scanning for changes and mounting/updating the datastore. During this transition, I think `multiple_host_access` will be false - but also at the same time, this is not really a stable thing to wait on, as there's too many variables that we would need to control (knowing that a disk is shared, or what hosts it exactly needs to be shared on, is currently out of the scope of this resource).

So I think this might be the simpler of the two scenarios. To that end I've updated things so that we specifically wait on the `ResourceInUse` error type instead of a string match (as per the other error matcher).